File: app/src/main/java/euphoria/psycho/browser/BrowserFragment.java
package euphoria.psycho.browser;

import android.annotation.SuppressLint;
import android.app.AlertDialog;
import android.app.Fragment;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.Bitmap;
import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.preference.PreferenceManager;
import android.text.TextUtils;
import android.view.*;
import android.webkit.*;
import android.widget.*;

import euphoria.psycho.browser.floating.FloatingActionButton;
import euphoria.psycho.browser.floating.FloatingActionsMenu;
import euphoria.psycho.common.Utils;

import java.io.*;
import java.lang.ref.WeakReference;
import java.text.Collator;
import java.util.*;
import java.util.stream.Collector;
import java.util.stream.Collectors;

public class BrowserFragment extends Fragment implements AdapterView.OnItemClickListener {

    private static final int CACHE_SIZE = 1024 * 1024 * 8;
    private static final String DEFAULT_URI = "https://m.youtube.com";
    private static final int INITIAL_PROGRESS = 5;
    private static final String KEY_FILTER = "filter";
    private static final String KEY_URI = "uri";
    private static final String KEY_IS_FILTER = "is_filter";
    private static final String PATTERN_URL = "m.youtube.com";
    private static final String SEARCH_MARK = "/*mark for changed*/";
    private String DEFAULT_FILTER;
    private String mCurrentUri = "";
    private String mFilters;
    private String mFilterSource;
    private Boolean mIsFilter = true;
    private WebView mWebView;
    private ProgressBar mProgressBar;
    private FloatingActionsMenu mFloating;
    private WeakReference<BottomSheet> mBottomSheet;

    private void applyFilter() {
        if (mIsFilter && mCurrentUri.contains(PATTERN_URL)) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
                mWebView.evaluateJavascript(mFilterSource, it -> {

                });
            }
        }
    }

    private void getFilterSource() {
        InputStream inputStream = getResources().openRawResource(R.raw.block_channel);
        try {
            BufferedReader reader = new BufferedReader(new InputStreamReader(inputStream, "UTF-8"));
            StringBuilder stringBuilder = new StringBuilder();
            if (TextUtils.isEmpty(mFilters)) {
                for (String line; (line = reader.readLine()) != null; ) {
                    stringBuilder.append(line).append('\n');
                }
            } else {
                for (String line; (line = reader.readLine()) != null; ) {
                    if (line.trim().startsWith(SEARCH_MARK)) {
                        stringBuilder.append(SEARCH_MARK).append('"').append(mFilters).append("\";\n");
                    } else {
                        stringBuilder.append(line).append('\n');
                    }
                }
            }
            mFilterSource = stringBuilder.toString();

        } catch (Exception e) {
            e.printStackTrace();
        }
        Utils.closeSilently(inputStream);
    }

    private void handleLongClick() {
        WebView.HitTestResult result = mWebView.getHitTestResult();
        switch (result.getType()) {
            case WebView.HitTestResult.SRC_ANCHOR_TYPE:
                showMenu(result.getExtra());
                break;
        }
    }

    private void loadUri() {
        mWebView.loadUrl(mCurrentUri);
    }

    // setting filters
    private void setFilter() {
        final EditText editText = new EditText(getActivity());

        final SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(getActivity());
        editText.setText(preferences.getString(KEY_FILTER, DEFAULT_FILTER));
        new AlertDialog.Builder(getActivity(), android.R.style.Theme_Holo_Dialog)
                .setView(editText)
                .setNegativeButton(android.R.string.cancel, ((dialog, which) -> {
                    dialog.dismiss();
                }))
                .setPositiveButton(android.R.string.ok, ((dialog, which) -> {
                    String filter = editText.getText().toString();
                    filter = sortFilters(filter);
                    preferences.edit().putString(KEY_FILTER, filter).apply();
                    mFilters = filter;
                    getFilterSource();

                    dialog.dismiss();
                }))
                .show();

    }

    // setting event handlers
    private void setupControls(View view) {
        if (mFloating == null) mFloating = view.findViewById(R.id.floating_menu);
        view.findViewById(R.id.floating_refresh).setOnClickListener(e -> {
            mWebView.reload();
            mFloating.collapse();
        });
        view.findViewById(R.id.floating_menu_button).setOnClickListener(e -> {
            List<BottomSheet.Item> items = new ArrayList<>();

            BottomSheet.Item item5 = new BottomSheet.Item();

            item5.title = "设置屏蔽频道";
            item5.imageResId = R.drawable.ic_filter_list_black_24dp;
            items.add(item5);


            BottomSheet.Item item3 = new BottomSheet.Item();

            if (mIsFilter) {
                item3.title = "开启屏蔽";
                item3.imageResId = R.drawable.ic_lock_outline_black_24dp;
            } else {
                item3.title = "关闭屏蔽";
                item3.imageResId = R.drawable.ic_lock_open_black_24dp;
            }
            items.add(item3);

            BottomSheet.Item item1 = new BottomSheet.Item();

            item1.title = "复制";
            item1.imageResId = R.drawable.ic_content_copy_black_24dp;
            items.add(item1);

            BottomSheet.Item item2 = new BottomSheet.Item();

            item2.title = "分享";
            item2.imageResId = R.drawable.ic_share_black_24dp;
            items.add(item2);

            BottomSheet.Item item6 = new BottomSheet.Item();

            item6.title = "返回";
            item6.imageResId = R.drawable.ic_chevron_left_black_24dp;
            items.add(item6);

            BottomSheet bottomSheet = new BottomSheet(getActivity(), items);
            mBottomSheet = new WeakReference<BottomSheet>(bottomSheet);
            bottomSheet.show(getView().findViewById(R.id.relative_layout), this);
            mFloating.collapse();
        });
        view.findViewById(R.id.floating_go_back).setOnClickListener(e -> {

            if (mWebView.canGoBack()) mWebView.goBack();
            mFloating.collapse();
        });
        view.findViewById(R.id.floating_home).setOnClickListener(e -> {
            mCurrentUri = DEFAULT_URI;
            loadUri();
            mFloating.collapse();
        });


    }

    @SuppressLint("SetJavaScriptEnabled")
    private void setupWebView() {
        mWebView.setOnLongClickListener(v ->
        {
            handleLongClick();
            return false;
        });

        mWebView.clearCache(false);
        mWebView.setLongClickable(true);
        mWebView.setWebChromeClient(new WebChromeClient() {
            @Override
            public void onProgressChanged(WebView view, int newProgress) {

                mProgressBar.setProgress(newProgress);
            }
        });
        mWebView.setWebViewClient(new WebViewClient() {
            @Override
            public void onPageFinished(WebView view, String url) {
                if (mProgressBar.getVisibility() == View.VISIBLE) {


                    mProgressBar.setVisibility(View.GONE);
                }
                applyFilter();
            }

            @Override
            public void onPageStarted(WebView view, String url, Bitmap favicon) {

                mProgressBar.setVisibility(View.VISIBLE);
                mProgressBar.setProgress(INITIAL_PROGRESS);
            }

            @Override
            public boolean shouldOverrideUrlLoading(WebView view, String url) {

                mCurrentUri = url;
                mWebView.loadUrl(url);
                return true;
            }
        });
        WebSettings settings = mWebView.getSettings();
        File cacheDirectory = new File(Environment.getExternalStorageDirectory(), ".cache_browser");
        if (!cacheDirectory.isDirectory()) {
            cacheDirectory.mkdir();
        }
        // Cache
        settings.setAppCacheEnabled(true);
        settings.setAppCachePath(cacheDirectory.getAbsolutePath());
        settings.setAppCacheMaxSize(CACHE_SIZE);
        settings.setCacheMode(WebSettings.LOAD_DEFAULT);

        // Storage & Database
        settings.setDomStorageEnabled(true);
        settings.setDatabaseEnabled(true);

        settings.setGeolocationEnabled(true);
        settings.setAllowFileAccess(true);

        settings.setJavaScriptEnabled(true);

        settings.setSupportZoom(false);
        settings.setDisplayZoomControls(false);
    }

    private void showMenu(String line) {
        View view = LayoutInflater.from(getActivity()).inflate(R.layout.menu_link, null);

        ((TextView) view.findViewById(R.id.text_link)).setText(line);

        final PopupWindow popupWindow = new PopupWindow(view, ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT, true);
        popupWindow.setTouchable(true);
        popupWindow.setFocusable(true);
        view.findViewById(R.id.text_copy_text).setOnClickListener(v -> {
            Utils.copyText(getActivity(), line);
            popupWindow.dismiss();
        });
        popupWindow.showAtLocation(getView().findViewById(R.id.relative_layout), Gravity.CENTER, 0, 0);
    }

    private void toggleFilter() {
        mIsFilter = !mIsFilter;
    }

    private static String sortFilters(String value) {
        Collator collator = Collator.getInstance(Locale.CHINA);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            return Arrays.stream(value.split("\\|"))
                    .sorted(collator::compare)
                    .distinct()
                    .collect(Collectors.joining("|"));

        } else {
            List<String> list = new ArrayList<>();

            for (String i : value.split("\\|")) {
                if (list.indexOf(i) > -1) continue;
                list.add(i);
            }
            Collections.sort(list, collator::compare);

            StringBuilder stringBuilder = new StringBuilder();
            for (String i : list) {
                stringBuilder.append(i).append('|');
            }
            stringBuilder.replace(stringBuilder.length() - 1, stringBuilder.length(), "");
            return stringBuilder.toString();
        }

    }


    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        setupWebView();
        loadUri();

    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
        

        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(getActivity());
        mCurrentUri = preferences.getString(KEY_URI, DEFAULT_URI);
        mFilters = preferences.getString(KEY_FILTER, DEFAULT_FILTER);

        if (Utils.isNullOrBlank(mFilters)) {
            List<String> list = Utils.readLines(getActivity(), R.raw.filter);
            if (!Utils.isEmpty(list)) {
                mFilters = Utils.joining(list, "|");
            }
        }

        mIsFilter = preferences.getBoolean(KEY_IS_FILTER, true);
        getFilterSource();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_browser, container, false);

        mWebView = view.findViewById(R.id.web_view);
        mProgressBar = view.findViewById(R.id.progress);
        setupControls(view);
        return view;
    }

    @Override
    public void onDestroy() {

        super.onDestroy();
    }

    @Override
    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {

        switch (position) {
            case 0:
                setFilter();
                break;
            case 1:
                toggleFilter();
                if (mIsFilter)
                    Toast.makeText(getActivity(), "开启屏蔽频道功能", Toast.LENGTH_LONG).show();
                else
                    Toast.makeText(getActivity(), "关闭屏蔽频道功能", Toast.LENGTH_LONG).show();

                break;
            case 2:
                Utils.copyText(getActivity(), mCurrentUri);
                Toast.makeText(getActivity(), "成功复制链接到剪切板", Toast.LENGTH_LONG).show();
                break;
            case 3:
                Intent intent = new Intent(Intent.ACTION_SEND);
                intent.setType("text/plain");
                intent.putExtra(Intent.EXTRA_TEXT, mCurrentUri);
                getActivity().startActivity(Intent.createChooser(intent, ""));
                break;
            default:
                break;
        }
        if (mBottomSheet.get() != null) {
            mBottomSheet.get().dismiss();
        }
    }

    @Override
    public void onPause() {
        super.onPause();
        PreferenceManager.getDefaultSharedPreferences(getActivity()).edit().putString(KEY_URI, mCurrentUri)
                .putBoolean(KEY_IS_FILTER, mIsFilter).apply();
    }

    @Override
    public void onResume() {
        super.onResume();

        mWebView.setOnKeyListener((v, keyCode, event) -> {
            if (event.getAction() == KeyEvent.ACTION_UP && keyCode == KeyEvent.KEYCODE_BACK) {
                if (mWebView.canGoBack()) {
                    mWebView.goBack();
                    return true;
                }
            }
            return false;
        });

    }

}

File: app/src/main/java/euphoria/psycho/browser/BrowserFragment.kt
package euphoria.psycho.browser

import android.graphics.Bitmap
import android.graphics.drawable.BitmapDrawable
import android.os.Build
import android.os.Bundle
import android.os.Environment
import android.view.*
import android.webkit.*
import android.widget.EditText
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.Fragment
import kotlinx.android.synthetic.main.fragment_browser.*
import kotlinx.android.synthetic.main.fragment_browser.view.*
import java.io.File
import java.nio.charset.Charset

class BrowserFragment : Fragment() {
    private var mCurrentUri = ""
    private var mFilters = ""
    private var mFilterSource = ""
    private var mIsFilter = true
    private var mIsFullScreen = false

    val applyFitler = {
        if (mIsFilter && mCurrentUri.contains("m.youtube.com")) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
                web_view?.evaluateJavascript(
                    mFilterSource
                ) {
                    requireContext().toast(it)
                }
            }
        }
    }
    val applySetFilterRule = {
        val editText = EditText(requireContext()).apply {
            setText(requireContext().sharedPreferences().string(KEY_FILTER))
        }
        AlertDialog.Builder(requireContext())
            .setView(editText)
            .setNegativeButton(android.R.string.cancel) { dialog, _ ->
                dialog.dismiss()
            }
            .show()
    }
    val setupSwipeLayout = {
        with(swipe_refresh) {
            setColorSchemeResources(R.color.colorAccent)
            setOnRefreshListener {
                reload()
            }
        }
        swipe_refresh.isRefreshing = false
    }
    val setUriTextBox = {
        text_url.run {
            setOnFocusChangeListener { v, hasFocus ->
                if (hasFocus) {

                }
            }
        }
    }
    val setupButtonGo = {
        image_go.setImageResource(R.drawable.ic_btn_reload)
    }
    val setupWebView = {
        web_view.let {
            it.clearCache(false)
            it.settings.run {
                val cachePath = File(Environment.getExternalStorageDirectory(), ".cache_browser")
                if (!cachePath.isDirectory) cachePath.mkdir()
                setAppCacheMaxSize(1024 * 1024 * 8)
                cacheMode = WebSettings.LOAD_DEFAULT
                domStorageEnabled = true
                databaseEnabled = true
                setAppCachePath(cachePath.absolutePath)
                javaScriptEnabled = true
                allowFileAccess = true
                setAppCacheEnabled(true)

            }
            it.isLongClickable = true
            it.setOnLongClickListener(linkHandler())
            it.webViewClient = getWebViewClient()

        }
    }
    val getFilterSource = {
        val inputStream = resources.openRawResource(R.raw.block_channel)
        val reader = inputStream.reader(Charset.forName("UTF-8"))
        if (mFilters.isNotBlank()) {
            val lines = reader.readLines()
            val stringBuilder = StringBuilder()
            for (i in lines.indices) {
                if (lines[i].trimStart().startsWith(SEARCH_MARK)) {
                    stringBuilder.append("$SEARCH_MARK\"$mFilters\";").append('\n')
                } else {
                    stringBuilder.append(lines[i]).append('\n')
                }
            }
            mFilterSource = stringBuilder.toString()
        } else {
            mFilterSource = reader.readText()
        }
        inputStream.close()
    }
    val getWebChromeClient: () -> WebChromeClient = {
        object : WebChromeClient() {
            override fun onProgressChanged(view: WebView?, newProgress: Int) {
                progress.progress = newProgress
            }

            override fun onShowCustomView(view: View?, callback: CustomViewCallback?) {
//                Log.e(TAG, "[onShowCustomView]")
//                mIsFullScreen = true
//                view?.let {
//                    browser_container.visibility = View.INVISIBLE
//                    val params = FrameLayout.LayoutParams(
//                        ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT
//                    )
//                } ?: run {
//                    appbar.setExpanded(false, true)
//                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
//                        web_view.isNestedScrollingEnabled = false
//                    }
//                }
//                switchToImmersiveMode()
            }
        }
    }
    val getWebViewClient: () -> WebViewClient = {
        object : WebViewClient() {
            override fun onPageStarted(view: WebView?, url: String?, favicon: Bitmap?) {

                favicon?.let {
                    text_url.setCompoundDrawables(BitmapDrawable(it), null, null, null)

                }
                progress.run {
                    progress = INITIAL_PROGRESS
                    visibility = View.VISIBLE
                }
            }

            override fun onLoadResource(view: WebView?, url: String?) {

                super.onLoadResource(view, url)
            }


            override fun onPageFinished(view: WebView?, url: String?) {
                text_url.setText(mCurrentUri)
                if (progress.visibility == View.VISIBLE) {
                    progress.visibility = View.GONE
                }
                swipe_refresh.isRefreshing = false
                applyFitler()

            }

            override fun shouldOverrideUrlLoading(view: WebView, request: WebResourceRequest): Boolean {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                    web_view.webChromeClient = getWebChromeClient()
                    view.loadUrl(request.url.toString())
                    return true
                } else return false
            }

            override fun onReceivedError(view: WebView?, request: WebResourceRequest?, error: WebResourceError?) {
                requireContext().toast(error.toString())
            }
        }
    }
    val linkHandler: () -> View.OnLongClickListener = {
        View.OnLongClickListener {
            val hitTestResult = web_view.hitTestResult
            when (hitTestResult.type) {
                WebView.HitTestResult.SRC_IMAGE_ANCHOR_TYPE -> {
                    onLongPress(hitTestResult.extra)
                }
            }
            false
        }
    }
    val loadUri = {
        web_view.loadUrl(mCurrentUri)
    }
    val onLongPress: (String?) -> Unit = {
    }
    val reload = {
        web_view.reload()
    }
    val switchToImmersiveMode = {
        requireActivity().window.run {
            addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
            decorView.systemUiVisibility = (View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                    or View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                    or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                    or View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
                    or View.SYSTEM_UI_FLAG_FULLSCREEN
                    or View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY)
        }
    }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
        val sharedPreferences = requireContext().sharedPreferences()
        mCurrentUri = sharedPreferences.string(KEY_URI, "https://m.youtube.com")
        mFilters = sharedPreferences.string(KEY_FILTER, DEFAULT_FILTER);
        getFilterSource()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_browser, container, false)
        return view
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_filter -> {
                applyFitler()
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        setupButtonGo()
        setupSwipeLayout()
        setupWebView()
        loadUri()
    }

    override fun onResume() {
        super.onResume()
        view?.run {
            isFocusableInTouchMode = true
            requestFocus()
            setOnKeyListener { v, keyCode, event ->
                if (event.getAction() == KeyEvent.ACTION_UP && keyCode == KeyEvent.KEYCODE_BACK) {
                    if (web_view.canGoBack()) {
                        web_view.goBack()

                        true;
                    } else {
                        false
                    }
                }
                false;
            }
        }
    }

    companion object {
        private const val TAG = "/BrowserFragment"
        private const val KEY_URI = "uri"
        private const val KEY_FILTER = "filter"
        private const val INITIAL_PROGRESS = 5
        private const val SEARCH_MARK = "/*mark for changed*/"
        private const val DEFAULT_FILTER =
            "54新觀點|从台湾看见世界|寰宇新聞 頻道|美国之音中文网|夢想街之全能事務所|民視新聞|明鏡火拍|年代向錢看|三立iNEWS|三立LIVE新聞|台視新聞 TTV NEWS|头条軍事【军事头条 軍情諜報 軍事解密 每日更新】歡迎訂閱|香港全城討論區|新聞面對面|新聞追追追"

    }

}